### ISSUE TITLE
Support Windows path manipulation on other platforms

### ISSUE THREAD

**Issue by mathstuf** [rust-lang/rust#60845]
It'd be nice if the facilities under `std::os` which don't use platform-specific APIs would be available on all platforms. For example, the ability to take a bytestream and construct a Windows path out of it in order to detect whether it is a valid Windows path would be great.

The use case is implementing a check which validates that a path is valid on Windows. The `OsStrExt` to get access to the underlying data as if it were a Windows path would be useful for this (since the check should be able to do its work on any platform).

I suspect that there are other APIs that would be useful for manipulating data as another platform might see it.

**Comment by omerbenamram** [comment]
I'm writing some cross platform forensics tools that need to manipulate windows paths on all platforms (joining, splitting, etc..).

In python I would use something like `ntpath` instead of `pathlib`, but it would be fantastic if rust could support that as well!

**Comment by retep998** [comment]
>the ability to take a bytestream and construct a Windows path out of it

Windows paths are not byte sequences. They are u16 sequences, which Rust converts into wtf-8.

**Comment by mathstuf** [comment]
Right, but what I get from `git ls-files` is a series of bytes. If the API requires that I batch them into u16 code units (with undefined endianess :/ ) or do some wtf-8 encoding process manually before passing to the "give me a Windows path" API, that's fine. It's understandable that the "here's an `A` path, give me a `W` path" is not implementable without Windows APIs due to code pages and the like.

**Comment by retep998** [comment]
The best you can hope for is getting utf-8 from git, but it won't support paths with lone surrogates. The worst case is git using the system narrow encoding on which case everything is awful.

**Comment by mathstuf** [comment]
From Git, yes (well, it actually supports arbitrary bytes, but that's part of what I'm trying to check and reject here). It obviously can't encode surrogate halves unambiguously anyways since it is basically byte arrays. However, there are other use cases for this. For example "is this path absolute?", "what is the directory separator?", or "please join these path components".

**Comment by RReverser** [comment]
I could use this too - my use case is JavaScript source map manipulation that should be able to deal with both Windows and Linux paths in sources, join paths, resolve them relatively one to another etc.

**Comment by sanbox-irl** [comment]
Are there no workarounds to this? Currently, I am managing my serialization with Strings (using serde), and then converting to a PathBuf at runtime and back near serialization time. It is not an excellent system, and has failed in multiple ways (forgetting to get the string prepared in time, especially). A possible solution would be a custom serde serializer/deserializer, but I'm not very confident in my ability to do that and it would be a sort of last ditch attempt

**Comment by mathstuf** [comment]
[This IRLO thread](https://internals.rust-lang.org/t/path-to-lexical-absolute/14940) may be of interest.
